File: course_materials/linguistics/python/plot.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
import cartopy.feature as cfeature
import seaborn as sns
import geopandas as gpd


def plot_ling_map(ling_data, qa_key, qid, point_size=1,
                  show_contiguous_us=True, ax=None):
    """
    Plot linguistic survey results on US map

    Parameters
    ----------
    ling_data : pd.DataFrame
        The linguistic data
    qa_key : pd.DataFrame
        The question and answer key
    qid : int
        The question ID to plot
    point_size : float, optional
        The size of the points on the map, by default 0.5
    show_contiguous_us : bool, optional
        Whether to show only the contiguous US, by default True
    ax : matplotlib.axes.Axes, optional
        The axes to plot on, by default None
    """

    # Format question column based on qid
    if qid < 100:
        qcol = f"Q0{qid}"
    else:
        qcol = f"Q{qid}"
    
    # Merge on answer_num and qcol
    qid_key = qa_key[qa_key['qid'] == qid]
    plt_df = ling_data.merge(qid_key, left_on=qcol, right_on='answer_num', how='left')
    
    # Filter out non-contiguous states if needed
    if show_contiguous_us:
        plt_df = plt_df[~plt_df['state_abb'].isin(['HI', 'AK'])]
    
    # Set up the map
    if ax is None:
        fig, ax = plt.subplots(figsize=(10, 8), subplot_kw={'projection': ccrs.PlateCarree()})
    
    # ax.add_feature(cfeature.BORDERS, linestyle=':')
    # ax.add_feature(cfeature.COASTLINE)
    ax.add_feature(cfeature.STATES, edgecolor='black')

    # Make colormap
    unique_answers = plt_df['answer'].unique()
    # palette = sns.color_palette("viridis", len(unique_answers))
    # color_map = dict(zip(unique_answers, palette))
    if len(unique_answers) <= 10:
        color_map = dict(zip(unique_answers, plt.cm.tab10.colors))
    else:
        color_map = dict(zip(unique_answers, plt.cm.tab20.colors))
    plt_df['color'] = plt_df['answer'].map(color_map)

    # Plot the data
    ax.scatter(plt_df['long'], plt_df['lat'], 
               c=plt_df['color'], s=point_size, 
               transform=ccrs.PlateCarree())
    
    # Add legend
    for answer, color in color_map.items():
        ax.scatter([], [], c=[color], label=answer)
    ax.legend(title="Answer", loc='center left', bbox_to_anchor=(1, 0.5))

    # Add title and labels
    ax.set_title(f"{qcol}: {qid_key['question'].values[0]}")

    return plt


def plot_ling_map_by_county(ling_data, qa_key, qid, 
                            linewidth=0.1, show_contiguous_us=True, ax=None):
    """
    Plot linguistic survey results, aggregated by county

    Parameters
    ----------
    ling_data : pd.DataFrame
        The linguistic survey data
    qa_key : pd.DataFrame
        The question and answer key
    qid : int
        The question ID
    linewidth : float, optional
        The width of the county borders, by default 0.1
    show_contiguous_us : bool, optional
        Whether to show only the contiguous US, by default True
    ax : matplotlib.axes.Axes, optional
        The axes to plot on, by default None
    """
    # Load county boundaries data from Geopandas
    map_df = gpd.read_file(
        'https://raw.githubusercontent.com/holtzy/The-Python-Graph-Gallery/master/static/data/US-counties.geojson'
    )
    state_df = pd.read_csv(
        'https://raw.githubusercontent.com/ChuckConnell/articles/refs/heads/master/fips2county.tsv',
        sep='\t',
        dtype={'StateFIPS': str}
    )[["StateFIPS", "StateAbbr"]].drop_duplicates()
    map_df = map_df.merge(state_df, how='left', left_on='STATE', right_on='StateFIPS')
    map_df["NAME"] = map_df["NAME"].str.lower()
    
    # Format question ID
    if qid < 100:
        qcol = f"Q0{qid}"
    else:
        qcol = f"Q{qid}"
    
    if show_contiguous_us:
        # Remove Hawaii and Alaska and Puerto Rico
        map_df = map_df[~map_df['StateAbbr'].isin(["HI", "AK"]) & ~map_df['StateAbbr'].isna()]
        ling_data = ling_data[~ling_data['state_abb'].isin(["HI", "AK"])]
    
    # Get responses for the question of interest
    qid_key = qa_key[qa_key['qid'] == qid]
    
    # Clean county names to match map data
    ling_data.loc[:, 'county'] = ling_data.loc[:, 'county'].str.replace(" parish", "").str.lower()
    
    # Merge with question key
    plt_df = ling_data.merge(qid_key, how='left', left_on=qcol, right_on='answer_num')
    plt_df = plt_df.dropna(subset=['answer'])
    
    # Aggregate responses by county
    plt_df = plt_df.groupby(['county', 'state_abb'])['answer'].agg(lambda x: x.value_counts().idxmax()).reset_index()
    
    # Merge with map data
    plt_df = map_df.merge(plt_df, how='left', left_on=['NAME', 'StateAbbr'], right_on=['county', 'state_abb'])
    
    # Plot the map
    if ax is None:
        fig, ax = plt.subplots(1, 1, figsize=(10, 6))
    plt_df.boundary.plot(ax=ax, linewidth=linewidth, color='black')
    plt_df.plot(column='answer', ax=ax, legend=True, edgecolor='black', linewidth=linewidth)
    leg = ax.get_legend()
    leg.set_bbox_to_anchor((1.05, 0.5))
    ax.set_title(f"{qcol}: {qid_key['question'].values[0]}")
    ax.axis('off')

    return plt


def plot_dr_scatter(X, components=[0, 1], color=None, 
                    point_size=1, point_alpha=1, color_label=None):
    """
    Plot dimension reduction results as scatter plot

    Parameters
    ----------
    X : np.ndarray
        The data matrix with scores to plot
    components : list, optional
        List of components to plot, by default [0, 1]
    color : np.ndarray, optional
        A vector to use for coloring data points, by default None
    point_size : float, optional
        The size of the points, by default 0.5
    point_alpha : float, optional
        The transparency of the points, by default 1
    """
    if len(components) == 2:
        # plot 2d scatter plot
        x_var = X.columns[components[0]]
        y_var = X.columns[components[1]]
        if color is None:
            plt.scatter(X[x_var], X[y_var], s=point_size)
        else:
            # if numeric, use viridis
            if np.issubdtype(color.dtype, np.number):
                plt.scatter(X[x_var], X[y_var], c=color, s=point_size, alpha=point_alpha, cmap='magma')
                plt.colorbar(label=color_label)
            else:
                # Make colormap
                unique_answers = np.unique(color)
                palette = sns.color_palette("magma", len(unique_answers))
                color_map = dict(zip(unique_answers, palette))
                color = [color_map[c] for c in color]
                plt.scatter(X[x_var], X[y_var], c=color, s=point_size, alpha=point_alpha)
                # Add legend
                for answer, color in color_map.items():
                    plt.scatter([], [], c=[color], label=answer)
                plt.legend(title=color_label, loc='center left', bbox_to_anchor=(1, 0.5))

        plt.xlabel(x_var)
        plt.ylabel(y_var)
    else:
        # plot pair plot
        sns.pairplot(X[X.columns[components]], hue=color, palette='magma', diag_kind='kde', plot_kws={'alpha': point_alpha, 's': point_size})

    plt.show()


def plot_dr_map(X, ling_data, components=[0, 1], 
                point_size=1, point_alpha=1,
                by_county=False, show_contiguous_us=True):
    """
    Plot dimension reduction results on US map

    Parameters
    ----------
    X : np.ndarray
        The data matrix with scores to plot
    ling_data : pd.DataFrame
        The linguistic data
    components : list, optional
        List of components to plot, by default [0, 1]
    point_size : float, optional
        The size of the points on the map, by default 0.5
    point_alpha : float, optional
        The transparency of the points, by default 1
    by_county : bool, optional
        Whether to plot by county, by default False
    show_contiguous_us : bool, optional
        Whether to show only the contiguous US, by default True
    """
    
    # Merge with ling_data
    ling_cols = [col for col in ling_data.columns if col in ["ZIP", "state_abb", "state", "city", "lat", "long", "county"]]
    plt_df = pd.concat([X, ling_data[ling_cols]], axis=1)
    
    # Filter out non-contiguous states if needed
    if show_contiguous_us:
        plt_df = plt_df[~plt_df['state_abb'].isin(['HI', 'AK'])]
    
    if by_county:
        # Load county boundaries data from Geopandas
        map_df = gpd.read_file(
            'https://raw.githubusercontent.com/holtzy/The-Python-Graph-Gallery/master/static/data/US-counties.geojson'
        )
        state_df = pd.read_csv(
            'https://raw.githubusercontent.com/ChuckConnell/articles/refs/heads/master/fips2county.tsv',
            sep='\t',
            dtype={'StateFIPS': str}
        )[["StateFIPS", "StateAbbr"]].drop_duplicates()
        map_df = map_df.merge(state_df, how='left', left_on='STATE', right_on='StateFIPS')
        map_df["NAME"] = map_df["NAME"].str.lower()
        if show_contiguous_us:
            map_df = map_df[~map_df['StateAbbr'].isin(["HI", "AK"]) & ~map_df['StateAbbr'].isna()]
        
        # Clean county names to match map data
        plt_df.loc[:, 'county'] = plt_df.loc[:, 'county'].str.replace(" parish", "").str.lower()
        
        # Aggregate responses by county
        plt_df = plt_df.groupby(['county', 'state_abb'])[X.columns].agg(lambda x: np.mean(x)).reset_index()
        
        # Merge with map data
        plt_df = map_df.merge(plt_df, how='left', left_on=['NAME', 'StateAbbr'], right_on=['county', 'state_abb'])
        
        # Plot the county map
        fig, ax = plt.subplots(len(components), figsize=(10, 6))
        for i, comp in enumerate(components):
            plt_df.boundary.plot(ax=ax[i], linewidth=0.1, color='black')
            plt_df.plot(column=X.columns[comp], ax=ax[i], legend=True, edgecolor='black', linewidth=0.1, cmap='magma')
            ax[i].set_title(X.columns[comp])
            ax[i].axis('off')

    else:
        # Plot the map
        fig, ax = plt.subplots(len(components), figsize=(10, 8), subplot_kw={'projection': ccrs.PlateCarree()})
        for i, comp in enumerate(components):
            ax[i].add_feature(cfeature.STATES, edgecolor='black')
            scatter = ax[i].scatter(plt_df['long'], plt_df['lat'], c=plt_df[X.columns[comp]], s=point_size, alpha=point_alpha, transform=ccrs.PlateCarree(), cmap='magma')
            cbar = plt.colorbar(scatter, ax=ax[i])
            cbar.set_label(X.columns[comp])

    plt.show()

File: course_materials/linguistics/R/plot.R
#' Plot linguistic survey results on US map
#'
#' @param ling_data the linguistic survey data
#' @param qa_key the question and answer key
#' @param qid the question ID
#' @param point_size the size of the points on the map
#' @param map_type the type of map to plot
#' @param show_contiguous_us whether to show only the contiguous US
#'
#' @returns a ggplot object
plot_ling_map <- function(ling_data, qa_key, qid, point_size = 0.5,
                          map_type = c("state", "county"),
                          viridis_option = "A",
                          show_contiguous_us = TRUE) {
  map_type <- match.arg(map_type)

  # get question ID
  qcol <- dplyr::case_when(
    qid < 100 ~ paste0("Q0", qid),
    TRUE ~ paste0("Q", qid)
  )
  merge_by <- "answer_num"
  names(merge_by) <- qcol

  # get responses for the question of interest
  qid_key <- qa_key |>
    dplyr::filter(qid == !!qid)

  # merge text responses with the survey data
  plt_df <- ling_data |>
    dplyr::left_join(qid_key, by = merge_by)

  if (show_contiguous_us) {
    # remove Hawaii and Alaska for plotting purposes
    plt_df <- plt_df |>
      dplyr::filter(!(state_abb %in% c("HI", "AK")))
  }

  # get state/county boundaries data
  map_df <- ggplot2::map_data(map_type)

  # make map plot
  plt <- plt_df |>
    ggplot2::ggplot() +
    ggplot2::geom_polygon(
      ggplot2::aes(x = long, y = lat, group = group),
      data = map_df, color = "black", fill = NA
    ) +
    ggplot2::geom_point(
      ggplot2::aes(x = long, y = lat, color = answer),
      size = point_size
    ) +
    ggplot2::labs(
      title = sprintf("%s: %s", qcol, qid_key$question[[1]]),
      color = "Answer"
    ) +
    ggplot2::theme(
      axis.title = ggplot2::element_blank(),
      axis.text = ggplot2::element_blank(),
      axis.ticks = ggplot2::element_blank(),
      panel.grid = ggplot2::element_blank(),
      panel.background = ggplot2::element_blank()
    ) +
    ggplot2::guides(
      color = ggplot2::guide_legend(override.aes = list(size = 3))
    ) +
    # scale it to make it look more like a map
    ggplot2::coord_fixed(1.5)

  return(plt)
}


#' Plot linguistic survey results, aggregated by county
#'
#' @param ling_data the linguistic survey data
#' @param qa_key the question and answer key
#' @param qid the question ID
#' @param linewidth the width of the county borders
#' @param show_contiguous_us whether to show only the contiguous US
#'
#' @returns a ggplot object
plot_ling_map_by_county <- function(ling_data, qa_key, qid, linewidth = 0.1,
                                    show_contiguous_us = TRUE) {

  # get county boundaries data
  map_df <- ggplot2::map_data("county")

  # get question ID
  qcol <- dplyr::case_when(
    qid < 100 ~ paste0("Q0", qid),
    TRUE ~ paste0("Q", qid)
  )
  merge_by <- "answer_num"
  names(merge_by) <- qcol

  if (show_contiguous_us) {
    # remove Hawaii and Alaska for plotting purposes
    ling_data <- ling_data |>
      dplyr::filter(!(state_abb %in% c("HI", "AK")))
  }

  # get responses for the question of interest
  qid_key <- qa_key |>
    dplyr::filter(qid == !!qid)

  # prepare data for plotting
  plt_df <- ling_data |>
    dplyr::left_join(qid_key, by = merge_by) |>
    # do some cleaning of the counties to match the built-in ggplot map data
    dplyr::mutate(
      county = dplyr::case_when(
        state == "louisiana" ~ stringr::str_remove(county, " parish"),
        county == "dekalb" ~ "de kalb",
        county == "desoto" ~ "de soto",
        (state == "virginia") & !(county %in% c("james city", "charles city")) ~
          stringr::str_remove(county, " city"),
        county == "dewitt" ~ "de witt",
        county == "dupage" ~ "du page",
        county == "lamoure" ~ "la moure",
        county == "laporte" ~ "la porte",
        county == "lasalle" ~ "la salle",
        county == "district of columbia" ~ "washington",
        TRUE ~ stringr::str_remove_all(county, "'")
      ),
      state = dplyr::case_when(
        state == "dc" ~ "district of columbia",
        TRUE ~ state
      )
    ) |>
    dplyr::filter(!is.na(answer)) |>
    # aggregate response by county
    dplyr::group_by(county, state) |>
    dplyr::summarise(
      answer = names(which.max(table(answer))[1]),
      .groups = "drop"
    ) |>
    # merge with county boundaries data
    dplyr::left_join(
      map_df, by = c("county" = "subregion", "state" = "region")
    )

  # look at county mismatches in merge
  # plt_df |>
  #   dplyr::filter(is.na(long)) |>
  #   dplyr::distinct(county, state)

  # make map plot
  plt <- plt_df |>
    ggplot2::ggplot() +
    ggplot2::geom_polygon(
      ggplot2::aes(x = long, y = lat, group = group, fill = answer),
      data = plt_df, color = "black", linewidth = linewidth
    ) +
    ggplot2::geom_polygon(
      ggplot2::aes(x = long, y = lat, group = group),
      data = map_df, color = "black", linewidth = linewidth, fill = NA
    ) +
    ggplot2::labs(
      title = sprintf("%s: %s", qcol, qid_key$question[[1]]),
      color = "Answer"
    ) +
    ggplot2::theme(
      axis.title = ggplot2::element_blank(),
      axis.text = ggplot2::element_blank(),
      axis.ticks = ggplot2::element_blank(),
      panel.grid = ggplot2::element_blank(),
      panel.background = ggplot2::element_blank()
    ) +
    ggplot2::guides(
      color = ggplot2::guide_legend(override.aes = list(size = 3))
    ) +
    # scale it to make it look more like a map
    ggplot2::coord_fixed(1.5)
  return(plt)
}


#' Plot dimension reduction results as scatter plot
#'
#' @param X the data matrix with scores to plot
#' @param components vector of components to plot
#' @param color a vector to use for coloring data points
#' @param point_size the size of the points
#' @param point_alpha the transparency of the points
#' @param viridis_option the viridis color option ("A" through "H")
#' @param ... additional arguments to pass to ggplot2::theme()
#'
#' @returns a ggplot object
plot_dr_scatter <- function(X, components = 1:2, color = NULL,
                            point_size = 0.5, point_alpha = 1,
                            viridis_option = "A", ...) {
  if (length(components) == 2) {
    # plot 2d scatter plot
    x_var <- names(X)[components[1]]
    y_var <- names(X)[components[2]]
    if (is.null(color)) {
      plt <- X |>
        ggplot2::ggplot() +
        ggplot2::aes(
          x = .data[[x_var]], y = .data[[y_var]]
        )
    } else {
      plt <- X |>
        dplyr::bind_cols(.color = color) |>
        ggplot2::ggplot() +
        ggplot2::aes(
          x = .data[[x_var]], y = .data[[y_var]], color = .color
        )
      if (is.numeric(color)) {
        plt <- plt +
          ggplot2::scale_color_viridis_c(option = viridis_option)
      }
    }
    plt <- plt +
      ggplot2::geom_point(size = point_size)
  } else {
    # plot pair plot
    plt <- ggwrappers::plot_pairs(
      X,
      columns = components,
      color_lower = color,
      point_size = point_size,
      point_alpha = point_alpha
    )
    if (is.numeric(color)) {
      plt <- plt +
        ggplot2::scale_color_viridis_c(option = viridis_option)
    }
  }
  plt <- plt +
    ggplot2::theme_minimal(...)
  return(plt)
}


#' Plot dimension reduction results on a map
#'
#' @param X the data matrix
#' @param ling_data the linguistic data
#' @param ndim the number of top components to plot
#' @param point_size the size of the points
#' @param point_alpha the transparency of the points
#' @param linewidth the width of the county borders
#' @param viridis_option the viridis color option ("A" through "H")
#' @param by_county whether to plot by county
#' @param show_contiguous_us whether to show only the contiguous US
#'
#' @returns a ggplot object
plot_dr_map <- function(X, ling_data, components = 1:2,
                        point_size = 0.5, point_alpha = 0.5, linewidth = 0.1,
                        viridis_option = "A",
                        by_county = FALSE, show_contiguous_us = TRUE) {
  # get state/county boundaries data
  map_type <- ifelse(by_county ,"county", "state")
  map_df <- ggplot2::map_data(map_type)
  state_map_df <- ggplot2::map_data("state")

  # get the data to plot
  X_df <- as.data.frame(X)
  plt_df <- ling_data |>
    dplyr::select(
      tidyselect::any_of(
        c("ZIP", "state_abb", "state", "city", "lat", "long", "county")
      )
    ) |>
    dplyr::bind_cols(X_df)
  if (show_contiguous_us) {
    # remove Hawaii and Alaska for plotting purposes
    plt_df <- plt_df |>
      dplyr::filter(!(state_abb %in% c("HI", "AK")))
  }
  if (by_county) {
    plt_df <- plt_df |>
      # do some cleaning of the counties to match the map data
      dplyr::mutate(
        county = dplyr::case_when(
          state == "louisiana" ~ stringr::str_remove(county, " parish"),
          county == "dekalb" ~ "de kalb",
          county == "desoto" ~ "de soto",
          (state == "virginia") & !(county %in% c("james city", "charles city")) ~
            stringr::str_remove(county, " city"),
          county == "dewitt" ~ "de witt",
          county == "dupage" ~ "du page",
          county == "lamoure" ~ "la moure",
          county == "laporte" ~ "la porte",
          county == "lasalle" ~ "la salle",
          county == "district of columbia" ~ "washington",
          TRUE ~ stringr::str_remove_all(county, "'")
        ),
        state = dplyr::case_when(
          state == "dc" ~ "district of columbia",
          TRUE ~ state
        )
      ) |>
      # aggregate results per county
      dplyr::group_by(county, state) |>
      dplyr::summarise(
        dplyr::across(
          tidyselect::all_of(colnames(X_df)), mean
        ),
        .groups = "drop"
      ) |>
      # merge with county boundaries data
      dplyr::left_join(
        map_df, by = c("county" = "subregion", "state" = "region")
      )
  }

  plt_ls <- list()
  for (i in components) {
    # make map plot
    if (by_county) {
      plt_ls[[i]] <- plt_df |>
        ggplot2::ggplot() +
        ggplot2::geom_polygon(
          ggplot2::aes(
            x = long, y = lat, group = group,
            fill = .data[[colnames(X_df)[i]]],
            color = .data[[colnames(X_df)[i]]]
          ),
          data = plt_df, linewidth = linewidth
        ) +
        ggplot2::geom_polygon(
          ggplot2::aes(x = long, y = lat, group = group),
          data = state_map_df,
          color = "black", linewidth = linewidth * 2, fill = NA
        )
    } else {
      plt_ls[[i]] <- plt_df |>
        ggplot2::ggplot() +
        ggplot2::geom_polygon(
          ggplot2::aes(x = long, y = lat, group = group),
          data = map_df, color = "black", fill = NA
        ) +
        ggplot2::geom_point(
          ggplot2::aes(
            x = long, y = lat, color = .data[[colnames(X_df)[i]]]
          ),
          size = point_size
        )
    }
    plt_ls[[i]] <- plt_ls[[i]] +
      ggplot2::scale_color_viridis_c(option = viridis_option) +
      ggplot2::scale_fill_viridis_c(option = viridis_option) +
      ggplot2::labs(
        title = sprintf("Component %s", i),
        color = "Value", fill = "Value"
      ) +
      ggplot2::theme(
        axis.title = ggplot2::element_blank(),
        axis.text = ggplot2::element_blank(),
        axis.ticks = ggplot2::element_blank(),
        panel.grid = ggplot2::element_blank(),
        panel.background = ggplot2::element_blank()
      ) +
      # scale it to make it look more like a map
      ggplot2::coord_fixed(1.5)
  }
  plt <- patchwork::wrap_plots(plt_ls)
  return(plt)
}
